docs/custom-filters.adoc

```diff
@@ -36,7 +51,176 @@ For common things like logging and metric facade APIs it is recommended to use t
+intercept all Fetch Responses. To support this case Kroxylicious provides an interfaces for all request types and
+response types supported by Kafka (at the version of Kafka Kroxylicious depends on). A filter implementation can
+implement any combination of these interfaces.
+
+There is no requirement that a Filter handles both the request and response halves of an RPC. A Filter can choose to
+intercept only the request, or only the response, or both the request and response.
+
+==== Examples
+
+To intercept all Fetch Requests your class would implement
+{filter-api-javadoc}/io/kroxylicious/proxy/filter/FetchRequestFilter.html[FetchRequestFilter]:
+
+[source,java]
+----
+public class FetchRequestClientIdFilter implements FetchRequestFilter {
+
+    @Override
+    public void onFetchRequest(short apiVersion,
+                               RequestHeaderData header,
+                               FetchRequestData request,
+                               KrpcFilterContext context) {
+        header.setClientId("fetch-client!");
+        context.forwardRequest(header, request);
+    }
+}
+----
+
+To intercept all Fetch Responses your class would implement
+{filter-api-javadoc}/io/kroxylicious/proxy/filter/FetchResponseFilter.html[FetchResponseFilter]:
+
+[source,java]
+----
+public class FetchRequestClientIdFilter implements FetchResponseFilter {
+
+    @Override
+    public void onFetchResponse(short apiVersion,
+                                ResponseHeaderData header,
+                                FetchResponseData response,
+                                KrpcFilterContext context) {
+        mutateResponse(response),
+        context.forwardResponse(header, response);
+    }
+    ...
+}
+----
+
+To intercept all Fetch Requests and all Fetch Responses your class would implement
+{filter-api-javadoc}/io/kroxylicious/proxy/filter/FetchRequestFilter.html[FetchRequestFilter] and
+{filter-api-javadoc}/io/kroxylicious/proxy/filter/FetchResponseFilter.html[FetchResponseFilter]:
+
+[source,java]
+----
+public class FetchRequestClientIdFilter implements FetchRequestFilter, FetchResponseFilter {
+
+    @Override
+    public void onFetchRequest(short apiVersion,
+                               RequestHeaderData header,
+                               FetchRequestData request,
+                               KrpcFilterContext context) {
+        header.setClientId("fetch-client!");
+        context.forwardRequest(header, request);
+    }
+
+    @Override
+    public void onFetchResponse(short apiVersion,
+                                ResponseHeaderData header,
+                                FetchResponseData response,
+                                KrpcFilterContext context) {
+        mutateResponse(response),
+        context.forwardResponse(header, response);
+    }
+}
+----
+Specific Message Filter interfaces are mutually exclusive with <<Request/Response Protocol Filters, Request/Response>> and
```

Comment:
Should we outline why they are mutually exclusive?

(Could be done in a separate PR)